File: Example/FWFramework/Classes/Module/Test/Module/TestBridgeController.m
//
//  TestBridgeController.m
//  FWFramework_Example
//
//  Created by wuyong on 2022/8/25.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

#import "TestBridgeController.h"

@interface TestJavascriptBridge : NSObject

@end

@implementation TestJavascriptBridge

+ (void)testObjcCallback:(WKWebView *)webView data:(id)data callback:(FWJsBridgeResponseCallback)callback
{
    NSLog(@"TestJavascriptBridge.testObjcCallback called: %@", data);
    callback(@"Response from TestJavascriptBridge.testObjcCallback");
}

@end

@implementation TestBridgeController

- (void)setupWebBridge:(FWWebViewJsBridge *)bridge {
    [FWWebViewJsBridge enableLogging];
    [bridge setErrorHandler:^(NSString *handlerName, id data, FWJsBridgeResponseCallback responseCallback) {
        [UIWindow fw_showMessageWithText:[NSString stringWithFormat:@"handler %@ undefined: %@", handlerName, data] style:FWToastStyleDefault completion:^{
            responseCallback(@"Response from errorHandler");
        }];
    }];
    [bridge setFilterHandler:^BOOL(NSString * _Nonnull handlerName, id  _Nonnull data, FWJsBridgeResponseCallback  _Nonnull responseCallback) {
        if ([handlerName isEqualToString:@"testFilterCallback"]) {
            NSLog(@"testFilterCallback called: %@", data);
            responseCallback(@"Response from testFilterCallback");
            return false;
        }
        
        return true;
    }];
    [bridge registerHandler:@"testObjcCallback" handler:^(id data, FWJsBridgeResponseCallback responseCallback) {
        NSLog(@"testObjcCallback called: %@", data);
        responseCallback(@"Response from testObjcCallback");
    }];
    [bridge registerClass:[TestJavascriptBridge class] package:nil context:nil withMapper:nil];
    
    NSLog(@"registeredHandlers: %@", [bridge getRegisteredHandlers]);
    [bridge callHandler:@"testJavascriptHandler" data:@{ @"foo":@"before ready" }];
}

- (void)callHandler:(id)sender {
    id data = @{ @"greetingFromObjC": @"Hi there, JS!" };
    [self.webView.fw_jsBridge callHandler:@"testJavascriptHandler" data:data responseCallback:^(id response) {
        NSLog(@"testJavascriptHandler responded: %@", response);
    }];
}

- (void)errorHandler:(id)sender {
    id data = @{ @"greetingFromObjC": @"Hi there, Error!" };
    [self.webView.fw_jsBridge callHandler:@"notFoundHandler" data:data responseCallback:^(id  _Nonnull responseData) {
        NSLog(@"notFoundHandler responded: %@", responseData);
    }];
}

- (void)filterHandler:(id)sender {
    id data = @{ @"greetingFromObjC": @"Hi there, Error!" };
    [self.webView.fw_jsBridge callHandler:@"testFilterHandler" data:data responseCallback:^(id  _Nonnull responseData) {
        NSLog(@"testFilterHandler responded: %@", responseData);
    }];
}

- (void)setupSubviews
{
    [self renderButtons:self.webView];
    
    self.requestUrl = [FWModuleBundle resourceURL:@"Bridge.html"].absoluteString;
}

- (void)renderButtons:(WKWebView*)webView {
    UIFont* font = [UIFont fontWithName:@"HelveticaNeue" size:12.0];
    CGFloat y = FWScreenHeight - FWTopBarHeight - UIScreen.fw_safeAreaInsets.bottom - 45;
    
    UIButton *callbackButton = [UIButton buttonWithType:UIButtonTypeRoundedRect];
    [callbackButton setTitle:@"Call" forState:UIControlStateNormal];
    [callbackButton addTarget:self action:@selector(callHandler:) forControlEvents:UIControlEventTouchUpInside];
    [self.view insertSubview:callbackButton aboveSubview:webView];
    callbackButton.frame = CGRectMake(10, y, 60, 35);
    callbackButton.titleLabel.font = font;
    
    UIButton *errorButton = [UIButton buttonWithType:UIButtonTypeRoundedRect];
    [errorButton setTitle:@"Error" forState:UIControlStateNormal];
    [errorButton addTarget:self action:@selector(errorHandler:) forControlEvents:UIControlEventTouchUpInside];
    [self.view insertSubview:errorButton aboveSubview:webView];
    errorButton.frame = CGRectMake(70, y, 60, 35);
    errorButton.titleLabel.font = font;
    
    UIButton *filterButton = [UIButton buttonWithType:UIButtonTypeRoundedRect];
    [filterButton setTitle:@"Filter" forState:UIControlStateNormal];
    [filterButton addTarget:self action:@selector(filterHandler:) forControlEvents:UIControlEventTouchUpInside];
    [self.view insertSubview:filterButton aboveSubview:webView];
    filterButton.frame = CGRectMake(130, y, 60, 35);
    filterButton.titleLabel.font = font;
    
    UIButton* reloadButton = [UIButton buttonWithType:UIButtonTypeRoundedRect];
    [reloadButton setTitle:@"Reload" forState:UIControlStateNormal];
    [reloadButton addTarget:webView action:@selector(reload) forControlEvents:UIControlEventTouchUpInside];
    [self.view insertSubview:reloadButton aboveSubview:webView];
    reloadButton.frame = CGRectMake(190, y, 60, 35);
    reloadButton.titleLabel.font = font;
    
    UIButton* jumpButton = [UIButton buttonWithType:UIButtonTypeRoundedRect];
    [jumpButton setTitle:@"Jump" forState:UIControlStateNormal];
    FWWeakifySelf();
    [jumpButton fw_addTouchBlock:^(id  _Nonnull sender) {
        FWStrongifySelf();
        self.webRequest = @"http://kvm.wuyong.site/jssdk.html";
    }];
    [self.view insertSubview:jumpButton aboveSubview:webView];
    jumpButton.frame = CGRectMake(250, y, 60, 35);
    jumpButton.titleLabel.font = font;
}

@end

File: Example/FWFramework/Classes/Module/Test/Module/TestBridgeController.swift
//
//  TestBridgeController.swift
//  FWFramework_Example
//
//  Created by wuyong on 2022/9/5.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import FWFramework
import WebKit

class TestJavascriptBridge: NSObject {
    @MainActor @objc static func testObjcCallbackDefaultBridge(_ context: WebViewJSBridge.Context) {
        print("TestJavascriptBridge.testObjcCallback called: \(context.parameters)")
        context.completion?("Response from TestJavascriptBridge.testObjcCallback")
    }
}

class TestBridgeController: WebController {
    private lazy var actionView: UIView = {
        let result = UIView()
        return result
    }()
    
    override func setupWebView() {
        super.setupWebView()
        
        webView.app.safeObserveProperty(\.canGoBack) { [weak self] webView, _ in
            self?.actionView.isHidden = webView.canGoBack
        }
        webView.app.jsBridgeEnabled = true
    }

    override func setupWebBridge(_ bridge: WebViewJSBridge) {
        bridge.isLogEnabled = true

        bridge.setErrorHandler { context in
            UIWindow.app.showMessage(text: "handler \(context.handlerName) undefined: \(context.parameters)", style: .default) {
                context.completion?("Response from errorHandler")
            }
        }

        bridge.setFilterHandler { context in
            if context.handlerName == "testFilterCallback" {
                print("testFilterCallback called: \(context.parameters)")
                context.completion?("Response from testFilterCallback")
                return false
            }
            return true
        }

        bridge.registerHandler("testObjcCallback") { context in
            print("testObjcCallback called: \(context.parameters)")
            context.completion?("Response from testObjcCallback")
        }

        bridge.registerClass(TestJavascriptBridge.self)
        print("registeredHandlers: \(bridge.getRegisteredHandlers())")
        bridge.callHandler("testJavascriptHandler", data: ["foo": "before ready"])
    }

    override func setupSubviews() {
        let font = UIFont.systemFont(ofSize: 12)
        let callbackButton = UIButton(type: .roundedRect)
        callbackButton.setTitle("Call", for: .normal)
        callbackButton.addTarget(self, action: #selector(callHandler(_:)), for: .touchUpInside)
        callbackButton.titleLabel?.font = font
        actionView.addSubview(callbackButton)
        callbackButton.layoutChain.size(width: 60, height: 35).left(10).bottom(10)

        let errorButton = UIButton(type: .roundedRect)
        errorButton.setTitle("Error", for: .normal)
        errorButton.addTarget(self, action: #selector(errorHandler(_:)), for: .touchUpInside)
        errorButton.titleLabel?.font = font
        actionView.addSubview(errorButton)
        errorButton.layoutChain.left(70).size(toView: callbackButton).bottom(toView: callbackButton)

        let filterButton = UIButton(type: .roundedRect)
        filterButton.setTitle("Filter", for: .normal)
        filterButton.addTarget(self, action: #selector(filterHandler(_:)), for: .touchUpInside)
        filterButton.titleLabel?.font = font
        actionView.addSubview(filterButton)
        filterButton.layoutChain.left(130).size(toView: callbackButton).bottom(toView: callbackButton)

        let reloadButton = UIButton(type: .roundedRect)
        reloadButton.setTitle("Reload", for: .normal)
        reloadButton.addTarget(webView, action: #selector(WKWebView.reload), for: .touchUpInside)
        reloadButton.titleLabel?.font = font
        actionView.addSubview(reloadButton)
        reloadButton.layoutChain.left(190).size(toView: callbackButton).bottom(toView: callbackButton)

        let jumpButton = UIButton(type: .roundedRect)
        jumpButton.setTitle("Jump", for: .normal)
        jumpButton.app.addTouch { [weak self] _ in
            self?.webRequest = "http://kvm.wuyong.site/jssdk.html"
        }
        jumpButton.titleLabel?.font = font
        actionView.addSubview(jumpButton)
        jumpButton.layoutChain.left(250).size(toView: callbackButton).bottom(toView: callbackButton)
    }

    override func setupLayout() {
        view.addSubview(actionView)
        actionView.layoutChain.horizontal(toSafeArea: .zero).bottom(toSafeArea: .zero).height(45)
        
        requestUrl = ModuleBundle.resourceURL("Bridge.html")?.absoluteString
    }

    @objc func callHandler(_ sender: Any) {
        let data = ["greetingFromObjC": "Hi there, JS!"]
        webView.app.jsBridge?.callHandler("testJavascriptHandler", data: data, callback: { response in
            print("testJavascriptHandler responded: \(APP.safeString(response))")
        })
    }

    @objc func errorHandler(_ sender: Any) {
        let data = ["greetingFromObjC": "Hi there, Error!"]
        webView.app.jsBridge?.callHandler("notFoundHandler", data: data, callback: { response in
            print("notFoundHandler responded: \(APP.safeString(response))")
        })
    }

    @objc func filterHandler(_ sender: Any) {
        let data = ["greetingFromObjC": "Hi there, Filter!"]
        webView.app.jsBridge?.callHandler("testFilterHandler", data: data, callback: { response in
            print("testFilterHandler responded: \(APP.safeString(response))")
        })
    }
}
